src/shared/utilities/timeoutUtils.ts

```diff
@@ -118,3 +135,43 @@ export async function waitUntil<T>(
         await new Promise(r => setTimeout(r, opt.interval))
     }
 }
+
+/**
+ * Utility function to race a Timeout object against a Promise. Handles Timeout expiration and cancellation,
+ * exposing access through the use of callbacks. Cleanup of the Timeout object is up to the caller since the
+ * Timeout may be used for more than one Promise. Remember to use `killTimer()` once all Promises have fulfilled.
+ *
+ * @param promise Promise that returns a generic type
+ * @param timeout Timeout token that will race against the promise
+ * @param opt.allowUndefined Output promise can resolve undefined (default: true)
+ * @param opt.onExpire Callback for when the promise expired. The callback can return a value
+ * @param opt.onCancel Callback for when the promise was cancelled. The callback can return a value
+ *
+ * @returns A Promise that returns if successful, or rejects when the Timeout was cancelled or expired.
+ */
+export function createTimedPromise<T>(
```

Comment:
This is essentially a different flavor of `waitUntil()`, isn't it? So naming it something like `waitUntilFoo` helps discoverability and helps devs understand/compare the options. It would be elucidating to say "the difference compared to waitUntil() is..." in the docstring.

Reply:
They are pretty similar, though the key difference is `createTimedPromise` uses a `Timeout` object and makes no attempts for retries (which can be done with `waitUntil` by setting the `timeout` to 0). At least for me, `waitUntil` has always been used as blocking call to try to get some output from the input function. `createTimePromise` instead returns a new Promise wrapping timeout functionality into the input `promise`. This seems to mainly be a naming issue: `waitUntil` implies that it should be used as a blocking call.

Reply:
> `createTimePromise` instead returns a new Promise wrapping timeout functionality into the input `promise`. This seems to mainly be a naming issue: `waitUntil` implies that it should be used as a blocking call.

True. But this is the case for all async functions. And although in this case, it might feel like it's more relevant than usual... is it really the _main_ purpose of this function? The main purpose is to call lambda after a `timeout`, no?

Reply:
Decided `waitTimeout` would be the best name.